### ISSUE TITLE
Regression in docker: panic in HTTP server when connecting via Unix socket

### ISSUE THREAD

**Issue by stefwalter** [cockpit-project/cockpit#3156]
Docker crashes which causes Cockpit's integration tests check-kubernetes to fail:

```
Nov 13 10:22:22 localhost.localdomain docker[973]: time="2015-11-13T04:22:22.292286912-05:00" level=info msg="GET /version"
Nov 13 10:22:22 localhost.localdomain docker[973]: 2015/11/13 04:22:22 http: panic serving @: runtime error: invalid memory address or nil pointer dereference
Nov 13 10:22:22 localhost.localdomain docker[973]: goroutine 33 [running]:
Nov 13 10:22:22 localhost.localdomain docker[973]: net/http.(*conn).serve.func1(0xc82012c6e0, 0x7fd28d861a88, 0xc8200245a0)
Nov 13 10:22:22 localhost.localdomain docker[973]: /usr/lib/golang/src/net/http/server.go:1287 +0xb5 fp=0xc8206a54c8 sp=0xc8206a53f8
Nov 13 10:22:22 localhost.localdomain docker[973]: runtime.call32(0x0, 0x1287be8, 0xc82006c170, 0x1800000018)
Nov 13 10:22:22 localhost.localdomain docker[973]: /usr/lib/golang/src/runtime/asm_amd64.s:437 +0x3e fp=0xc8206a54f0 sp=0xc8206a54c8
Nov 13 10:22:22 localhost.localdomain docker[973]: runtime.gopanic(0xf0f880, 0xc82000e070)
Nov 13 10:22:22 localhost.localdomain docker[973]: /usr/lib/golang/src/runtime/panic.go:423 +0x4e9 fp=0xc8206a5570 sp=0xc8206a54f0
Nov 13 10:22:22 localhost.localdomain docker[973]: runtime.panicmem()
Nov 13 10:22:22 localhost.localdomain docker[973]: /usr/lib/golang/src/runtime/panic.go:42 +0x49 fp=0xc8206a5598 sp=0xc8206a5570
Nov 13 10:22:22 localhost.localdomain docker[973]: runtime.sigpanic()
Nov 13 10:22:22 localhost.localdomain docker[973]: /usr/lib/golang/src/runtime/sigpanic_unix.go:24 +0x2ba fp=0xc8206a55e8 sp=0xc8206a5598
Nov 13 10:22:22 localhost.localdomain docker[973]: github.com/docker/docker/api/server.getpwuid(0xffffffff, 0x0, 0x0, 0x0, 0x0)
Nov 13 10:22:22 localhost.localdomain docker[973]: /builddir/build/BUILD/docker-28c300fafb58c380d78381e08e1be35dfed5d4f9/_build/src/github.com/docker/docker/api/server/credentials_linux.go:75 +0x1f6 fp=0xc8206a5678 sp=0xc8206a55e8
Nov 13 10:22:22 localhost.localdomain docker[973]: github.com/docker/docker/api/server.(*Server).LogAction(0xc820190380, 0x7fd28d86f4e8, 0xc82012c790, 0xc820084620, 0x0, 0x0)
Nov 13 10:22:22 localhost.localdomain docker[973]: /builddir/build/BUILD/docker-28c300fafb58c380d78381e08e1be35dfed5d4f9/_build/src/github.com/docker/docker/api/server/credentials_linux.go:172 +0xaac fp=0xc8206a5850 sp=0xc8206a5678
Nov 13 10:22:22 localhost.localdomain docker[973]: github.com/docker/docker/api/server.(*Server).makeHttpHandler.func1(0x7fd28d86f4e8, 0xc82012c790, 0xc820084620)
Nov 13 10:22:22 localhost.localdomain docker[973]: /builddir/build/BUILD/docker-28c300fafb58c380d78381e08e1be35dfed5d4f9/_build/src/github.com/docker/docker/api/server/server.go:1680 +0x8aa fp=0xc8206a5aa8 sp=0xc8206a5850
Nov 13 10:22:22 localhost.localdomain docker[973]: net/http.HandlerFunc.ServeHTTP(0xc82032c300, 0x7fd28d86f4e8, 0xc82012c790, 0xc820084620)
Nov 13 10:22:22 localhost.localdomain docker[973]: /usr/lib/golang/src/net/http/server.go:1422 +0x3a fp=0xc8206a5ac8 sp=0xc8206a5aa8
Nov 13 10:22:22 localhost.localdomain docker[973]: github.com/gorilla/mux.(*Router).ServeHTTP(0xc8202019a0, 0x7fd28d86f4e8, 0xc82012c790, 0xc820084620)
Nov 13 10:22:22 localhost.localdomain docker[973]: /builddir/build/BUILD/docker-28c300fafb58c380d78381e08e1be35dfed5d4f9/vendor/src/github.com/gorilla/mux/mux.go:98 +0x29e fp=0xc8206a5be0 sp=0xc8206a5ac8
Nov 13 10:22:22 localhost.localdomain docker[973]: net/http.serverHandler.ServeHTTP(0xc820388120, 0x7fd28d86f4e8, 0xc82012c790, 0xc820084620)
Nov 13 10:22:22 localhost.localdomain docker[973]: /usr/lib/golang/src/net/http/server.go:1862 +0x19e fp=0xc8206a5c40 sp=0xc8206a5be0
Nov 13 10:22:22 localhost.localdomain docker[973]: net/http.(*conn).serve(0xc82012c6e0)
Nov 13 10:22:22 localhost.localdomain docker[973]: /usr/lib/golang/src/net/http/server.go:1361 +0xbee fp=0xc8206a5f98 sp=0xc8206a5c40
Nov 13 10:22:22 localhost.localdomain docker[973]: runtime.goexit()
Nov 13 10:22:22 localhost.localdomain docker[973]: /usr/lib/golang/src/runtime/asm_amd64.s:1696 +0x1 fp=0xc8206a5fa0 sp=0xc8206a5f98
Nov 13 10:22:22 localhost.localdomain docker[973]: created by net/http.(*Server).Serve
Nov 13 10:22:22 localhost.localdomain docker[973]: /usr/lib/golang/src/net/http/server.go:1910 +0x3f6
Nov 13 10:22:22 localhost.localdomain docker[973]: time="2015-11-13T04:22:22.394599607-05:00" level=info msg="GET /version"
Nov 13 10:22:22 localhost.localdomain docker[973]: 2015/11/13 04:22:22 http: panic serving @: runtime error: invalid memory address or nil pointer dereference
Nov 13 10:22:22 localhost.localdomain docker[973]: goroutine 34 [running]:
Nov 13 10:22:22 localhost.localdomain docker[973]: net/http.(*conn).serve.func1(0xc82012c840, 0x7fd28d861a88, 0xc8200245c8)
Nov 13 10:22:22 localhost.localdomain docker[973]: /usr/lib/golang/src/net/http/server.go:1287 +0xb5 fp=0xc8206a54c8 sp=0xc8206a53f8
Nov 13 10:22:22 localhost.localdomain docker[973]: runtime.call32(0x0, 0x1287be8, 0xc82006c170, 0x1800000018)
Nov 13 10:22:22 localhost.localdomain docker[973]: /usr/lib/golang/src/runtime/asm_amd64.s:437 +0x3e fp=0xc8206a54f0 sp=0xc8206a54c8
Nov 13 10:22:22 localhost.localdomain docker[973]: runtime.gopanic(0xf0f880, 0xc82000e070)
Nov 13 10:22:22 localhost.localdomain docker[973]: /usr/lib/golang/src/runtime/panic.go:423 +0x4e9 fp=0xc8206a5570 sp=0xc8206a54f0
Nov 13 10:22:22 localhost.localdomain docker[973]: runtime.panicmem()
Nov 13 10:22:22 localhost.localdomain docker[973]: /usr/lib/golang/src/runtime/panic.go:42 +0x49 fp=0xc8206a5598 sp=0xc8206a5570
Nov 13 10:22:22 localhost.localdomain docker[973]: runtime.sigpanic()
Nov 13 10:22:22 localhost.localdomain docker[973]: /usr/lib/golang/src/runtime/sigpanic_unix.go:24 +0x2ba fp=0xc8206a55e8 sp=0xc8206a5598
Nov 13 10:22:22 localhost.localdomain docker[973]: github.com/docker/docker/api/server.getpwuid(0xffffffff, 0x0, 0x0, 0x0, 0x0)
Nov 13 10:22:22 localhost.localdomain docker[973]: /builddir/build/BUILD/docker-28c300fafb58c380d78381e08e1be35dfed5d4f9/_build/src/github.com/docker/docker/api/server/credentials_linux.go:75 +0x1f6 fp=0xc8206a5678 sp=0xc8206a55e8
Nov 13 10:22:22 localhost.localdomain docker[973]: github.com/docker/docker/api/server.(*Server).LogAction(0xc820190380, 0x7fd28d86f4e8, 0xc82012c8f0, 0xc820084700, 0x0, 0x0)
Nov 13 10:22:22 localhost.localdomain docker[973]: /builddir/build/BUILD/docker-28c300fafb58c380d78381e08e1be35dfed5d4f9/_build/src/github.com/docker/docker/api/server/credentials_linux.go:172 +0xaac fp=0xc8206a5850 sp=0xc8206a5678
Nov 13 10:22:22 localhost.localdomain docker[973]: github.com/docker/docker/api/server.(*Server).makeHttpHandler.func1(0x7fd28d86f4e8, 0xc82012c8f0, 0xc820084700)
Nov 13 10:22:22 localhost.localdomain docker[973]: /builddir/build/BUILD/docker-28c300fafb58c380d78381e08e1be35dfed5d4f9/_build/src/github.com/docker/docker/api/server/server.go:1680 +0x8aa fp=0xc8206a5aa8 sp=0xc8206a5850
Nov 13 10:22:22 localhost.localdomain docker[973]: net/http.HandlerFunc.ServeHTTP(0xc82032c300, 0x7fd28d86f4e8, 0xc82012c8f0, 0xc820084700)
Nov 13 10:22:22 localhost.localdomain docker[973]: /usr/lib/golang/src/net/http/server.go:1422 +0x3a fp=0xc8206a5ac8 sp=0xc8206a5aa8
Nov 13 10:22:22 localhost.localdomain docker[973]: github.com/gorilla/mux.(*Router).ServeHTTP(0xc8202019a0, 0x7fd28d86f4e8, 0xc82012c8f0, 0xc820084700)
Nov 13 10:22:22 localhost.localdomain kubelet[1070]: E1113 04:22:22.394372    1070 event.go:194] Unable to write event: 'Post http://127.0.0.1:8080/api/v1/namespaces/default/events: dial tcp 127.0.0.1:8080: getsockopt: connection refused' (may retry after sleeping)
Nov 13 10:22:22 localhost.localdomain docker[973]: /builddir/build/BUILD/docker-28c300fafb58c380d78381e08e1be35dfed5d4f9/vendor/src/github.com/gorilla/mux/mux.go:98 +0x29e fp=0xc8206a5be0 sp=0xc8206a5ac8
Nov 13 10:22:22 localhost.localdomain docker[973]: net/http.serverHandler.ServeHTTP(0xc820388120, 0x7fd28d86f4e8, 0xc82012c8f0, 0xc820084700)
Nov 13 10:22:22 localhost.localdomain docker[973]: /usr/lib/golang/src/net/http/server.go:1862 +0x19e fp=0xc8206a5c40 sp=0xc8206a5be0
Nov 13 10:22:22 localhost.localdomain docker[973]: net/http.(*conn).serve(0xc82012c840)
Nov 13 10:22:22 localhost.localdomain docker[973]: /usr/lib/golang/src/net/http/server.go:1361 +0xbee fp=0xc8206a5f98 sp=0xc8206a5c40
Nov 13 10:22:22 localhost.localdomain docker[973]: runtime.goexit()
Nov 13 10:22:22 localhost.localdomain docker[973]: /usr/lib/golang/src/runtime/asm_amd64.s:1696 +0x1 fp=0xc8206a5fa0 sp=0xc8206a5f98
Nov 13 10:22:22 localhost.localdomain docker[973]: created by net/http.(*Server).Serve
Nov 13 10:22:22 localhost.localdomain docker[973]: /usr/lib/golang/src/net/http/server.go:1910 +0x3f6
```

Fedora 23 bug: https://bugzilla.redhat.com/show_bug.cgi?id=1275593
Upstream bug: https://github.com/rhatdan/docker/pull/149

CC @rhatdan @nalind

**Comment by stefwalter** [comment]
Ooops, it happened again
```
Traceback (most recent call last):
  File "./check-kubernetes", line 285, in testTopology
    b.wait_text("#service-list tr[data-name='mock'] td.containers", "1")
  File "/data/src/cockpit/test/testlib.py", line 247, in wait_text
    return self.wait_js_func('ph_text_is', selector, text)
  File "/data/src/cockpit/test/testlib.py", line 211, in wait_js_func
    return self.phantom.wait("%s(%s)" % (func, ','.join(map(jsquote, args))))
  File "/data/src/cockpit/test/testlib.py", line 670, in <lambda>
    return lambda *args: self._invoke(name, *args)
  File "/data/src/cockpit/test/testlib.py", line 693, in _invoke
    raise Error(res['error'])
Error: timeout
```

**Comment by stefwalter** [comment]
Version of docker is: docker-1.8.2-10.git28c300f.fc23.x86_64

**Comment by stefwalter** [comment]
Ooops, it happened again
```
Traceback (most recent call last):
  File "check-kubernetes", line 162, in testDashboard
    b.wait_text("#service-list tr[data-name='mock']:first-of-type td.containers", "1")
  File "testlib.py", line 247, in wait_text
    return self.wait_js_func('ph_text_is', selector, text)
  File "testlib.py", line 211, in wait_js_func
    return self.phantom.wait("%s(%s)" % (func, ','.join(map(jsquote, args))))
  File "testlib.py", line 670, in <lambda>
    return lambda *args: self._invoke(name, *args)
  File "testlib.py", line 693, in _invoke
    raise Error(res['error'])
Error: timeout
```

**Comment by stefwalter** [comment]
The broken Fedora 23 update: https://bodhi.fedoraproject.org/updates/FEDORA-2015-7e1a61e141

**Comment by stefwalter** [comment]
CC @lsm5

**Comment by stefwalter** [comment]
Ooops, it happened again
```
Traceback (most recent call last):
  File "check-kubernetes", line 285, in testTopology
    b.wait_text("#service-list tr[data-name='mock'] td.containers", "1")
  File "testlib.py", line 247, in wait_text
    return self.wait_js_func('ph_text_is', selector, text)
  File "testlib.py", line 211, in wait_js_func
    return self.phantom.wait("%s(%s)" % (func, ','.join(map(jsquote, args))))
  File "testlib.py", line 670, in <lambda>
    return lambda *args: self._invoke(name, *args)
  File "testlib.py", line 693, in _invoke
    raise Error(res['error'])
Error: timeout
```

**Comment by cockpituous** [comment]
Ooops, it happened again
```
Traceback (most recent call last):
  File "check-kubernetes", line 162, in testDashboard
    b.wait_text("#service-list tr[data-name='mock']:first-of-type td.containers", "1")
  File "/build/cockpit/test/testlib.py", line 248, in wait_text
    return self.wait_js_func('ph_text_is', selector, text)
  File "/build/cockpit/test/testlib.py", line 212, in wait_js_func
    return self.phantom.wait("%s(%s)" % (func, ','.join(map(jsquote, args))))
  File "/build/cockpit/test/testlib.py", line 674, in <lambda>
    return lambda *args: self._invoke(name, *args)
  File "/build/cockpit/test/testlib.py", line 697, in _invoke
    raise Error(res['error'])
Error: timeout
```
http://files.cockpit-project.org/logs/pull-3138-d3ec9a67-fedora-23-x86_64/log.html

**Comment by cockpituous** [comment]
Ooops, it happened again
```
Traceback (most recent call last):
  File "check-kubernetes", line 162, in testDashboard
    b.wait_text("#service-list tr[data-name='mock']:first-of-type td.containers", "1")
  File "/build/cockpit/test/testlib.py", line 248, in wait_text
    return self.wait_js_func('ph_text_is', selector, text)
  File "/build/cockpit/test/testlib.py", line 212, in wait_js_func
    return self.phantom.wait("%s(%s)" % (func, ','.join(map(jsquote, args))))
  File "/build/cockpit/test/testlib.py", line 674, in <lambda>
    return lambda *args: self._invoke(name, *args)
  File "/build/cockpit/test/testlib.py", line 697, in _invoke
    raise Error(res['error'])
Error: timeout
```
http://files.cockpit-project.org/logs/pull-3138-9fd74295-fedora-23-x86_64/log.html

**Comment by cockpituous** [comment]
Ooops, it happened again
```
Traceback (most recent call last):
  File "check-kubernetes", line 162, in testDashboard
    b.wait_text("#service-list tr[data-name='mock']:first-of-type td.containers", "1")
  File "/build/cockpit/test/testlib.py", line 248, in wait_text
    return self.wait_js_func('ph_text_is', selector, text)
  File "/build/cockpit/test/testlib.py", line 212, in wait_js_func
    return self.phantom.wait("%s(%s)" % (func, ','.join(map(jsquote, args))))
  File "/build/cockpit/test/testlib.py", line 674, in <lambda>
    return lambda *args: self._invoke(name, *args)
  File "/build/cockpit/test/testlib.py", line 697, in _invoke
    raise Error(res['error'])
Error: timeout
```
http://files.cockpit-project.org/logs/pull-3138-3d185dae-fedora-23-x86_64/log.html

**Comment by cockpituous** [comment]
Ooops, it happened again
```
Traceback (most recent call last):
  File "check-kubernetes", line 162, in testDashboard
    b.wait_text("#service-list tr[data-name='mock']:first-of-type td.containers", "1")
  File "testlib.py", line 248, in wait_text
    return self.wait_js_func('ph_text_is', selector, text)
  File "testlib.py", line 212, in wait_js_func
    return self.phantom.wait("%s(%s)" % (func, ','.join(map(jsquote, args))))
  File "testlib.py", line 674, in <lambda>
    return lambda *args: self._invoke(name, *args)
  File "testlib.py", line 697, in _invoke
    raise Error(res['error'])
Error: timeout
```
http://files.cockpit-project.org/logs/pull-3138-a458154d-fedora-23-x86_64/log.html

**Comment by stefwalter** [comment]
A new docker included in new images from #3138 should solve this problem.
